### PR TITLE
add new layout D_X,E_X for BB51.

### LINKED ISSUE #207: D_X_* & E_X_* Layout
* [x] Validate that other targets are not affected
* [x] Validate that new targets are build
* [x] Validate Bluejay D functionality
* [x]  Validate Bluejay E functionality
* [x] Validate BLHeli_S D functionality
* [x] Validate BLHeli_S E functionality
* [x] Validate BLHeli_S D flashing
* [x] Validate BLHeli_S E flashing

### PR DESCRIPTION
**Pull request by HAKRC**
I've added D_X and E_X layouts for BB51, and E_X_40_24 and E_X_40_48 have been validated, but D_X has not yet been validated.

### CODE REVIEW

**Comment by stylesuxx** [comment]
Thank you for your contribution! In our manufacturer discord we asked - quite some time ago - if anyone uses those layouts. And if so, if we could get samples.

We do not want to clutter the layouts with untested, unverified layouts - that's why we have not added them yet.

Are you going to use (and verify) the D Layout too?

Any chance you could provide some samples for each layout? Would be great to have something as reference just in case users run into problems so we have some hardware to test against.

**Comment by HAKRC** [comment]
Hello, we are already using it, we cooperate with BLS with D_X and E_X. Please give me an address, I send the sample to the past! Thanks!









 ----------Reply to Message----------
 On Fri, Jul 19, 2024 10:43 AM Chris ***@***.***&gt; wrote:

  

 
Thank you for your contribution! In our manufacturer discord we asked - quite some time ago - if anyone uses those layouts. And if so, if we could get samples.
 
We do not want to clutter the layouts with untested, unverified layouts - that's why we have not added them yet.
 
Are you going to use (and verify) the D Layout too?
 
Any chance you could provide some samples for each layout? Would be great to have something as reference just in case users run into problems so we have some hardware to test against.

**Comment by stylesuxx** [comment]
Hey, I pinged you (Hackrc Wang) on discord in our manufacturer chat, let's continue discussion there.

File: src/Bluejay.asm
Comment (hyp0dermik-code) [inline src/Bluejay.asm:1]: This file should not be changed

Comment (qq895657030) [reply src/Bluejay.asm:1]: We have changed the DEFAULT_PGM_STARTUP_POWER_MAX to increase the motor startup power limit, ensuring smoother motor startups. DEFAULT_PGM_STARTUP_POWER_MAX=5 is too low, and the motor sometimes fails to start.

Comment (HAKRC) [reply src/Bluejay.asm:1]: We have changed the DEFAULT_PGM_STARTUP_POWER_MAX to increase the motor startup power limit, ensuring smoother motor startups. DEFAULT_PGM_STARTUP_POWER_MAX=5 is too low, and the motor sometimes fails to start.

Comment (hyp0dermik-code) [reply src/Bluejay.asm:1]: You will have to set this as a custom setting in your build - see https://github.com/bird-sanctuary/bluejay/wiki/Development#building-from-source-with-custom-defaults

This change will not be accepted as this affects the defaults for all layouts

Comment (HAKRC) [reply src/Bluejay.asm:1]: OK,I made a new commit and restored the default settings in Bluejay.asm.

File: src/Layouts/Base.inc [hunk 1]
```diff
@@ -133,7 +133,7 @@ Set_Comparator_Phase_A MACRO
 IF MCU_TYPE == MCU_BB1 or MCU_TYPE == MCU_BB2
     mov  CMP_MX, #((A_Mux SHL 4) + V_Mux)
 ELSEIF MCU_TYPE == MCU_BB51
-    mov  CMP_MX, #12h
+    mov  CMP_MX, #(10h+A_Mux-1)
```
Comment (stylesuxx) [inline src/Layouts/Base.inc]: Spaces missing before and after `+` & `-` are missing.

File: src/Layouts/Base.inc [hunk 2]
```diff
@@ -142,7 +142,7 @@ Set_Comparator_Phase_B MACRO
 IF MCU_TYPE == MCU_BB1 or MCU_TYPE == MCU_BB2
     mov  CMP_MX, #((B_Mux SHL 4) + V_Mux)
 ELSEIF MCU_TYPE == MCU_BB51
-    mov  CMP_MX, #10h
+	mov  CMP_MX, #(10h+B_Mux-1)
```
Comment (stylesuxx) [inline src/Layouts/Base.inc]: Spaces missing before and after `+` & `-` are missing.

File: src/Layouts/Base.inc [hunk 3]
```diff
@@ -151,7 +151,7 @@ Set_Comparator_Phase_C MACRO
 IF MCU_TYPE == MCU_BB1 or MCU_TYPE == MCU_BB2
     mov  CMP_MX, #((C_Mux SHL 4) + V_Mux)
 ELSEIF MCU_TYPE == MCU_BB51
-    mov  CMP_MX, #11h
+    mov  CMP_MX, #(10h+C_Mux-1)
```
Comment (stylesuxx) [inline src/Layouts/Base.inc]: Spaces missing before and after `+` & `-` are missing.

**Review by stylesuxx — CHANGES_REQUESTED** [review]
Just some minor formatting nitpicks.

File: src/Layouts/Base.inc (hunk 1 above)
Comment (HAKRC) [reply src/Layouts/Base.inc]: The format has been adjusted.

**Comment by HAKRC** [comment]
> Just some minor formatting nitpicks.

The format has been adjusted.

**Comment by hyp0dermik-code** [comment]
The new layouts need to be added to Bluejay.asm after L118:
https://github.com/HAKRC/bluejay/blob/bed3edb9e378fa559baf4541a8214e63d01b48ff/src/Bluejay.asm#L115

**Comment by hyp0dermik-code** [comment]
The Makefile needs to be altered to include the new layouts too:
```
LAYOUTS_X               = A B C D E
```

**Comment by HAKRC** [comment]
> The Makefile needs to be altered to include the new layouts too:
> 
> ```
> LAYOUTS_X               = A B C D E
> ```

Sure,Makefile has updated.

**Comment by HAKRC** [comment]
We have sent you the boards with D_X and E_X layouts. Have you received them?

**Comment by HAKRC** [comment]
"We hope the latest firmware is updated to https://esc-configurator.com/, so our customers can update the code and configuration there."

**Review by stylesuxx — APPROVED** [review]
Looking good so far. Bench tested D&E layout. Defaults come flashed with BLHeli_S. Tested stock, flashed Bluejay, made sure power consumtion is in a similar range and motors spin smoothly. RPM is reported in Betaflight configurator.

Also see #207

Configurator needs to be re-deployed [after this change](https://github.com/stylesuxx/esc-configurator/issues/481).

**Comment by HAKRC** [comment]
Can I ask about the release date of the new version?

**Comment by stylesuxx** [comment]
We are planning on releasing a RC later today.

**Comment by hyp0dermik-code** [comment]
@HAKRC the RC has been released and can be downloaded via https://esc-configurator.com by enabling **Show firmware pre-releases** in the **Settings** of https://esc-configurator.com 
image

If you do not see this, make sure your esc-configurator.com is up to date, either by the popup or force-refreshing the page.

**Comment by HAKRC** [comment]
> We are planning on releasing a RC later today.

We appreciate it .

**Comment by HAKRC** [comment]
> @HAKRC the RC has been released and can be downloaded via https://esc-configurator.com by enabling **Show firmware pre-releases** in the **Settings** of https://esc-configurator.com image
> 
> If you do not see this, make sure your esc-configurator.com is up to date, either by the popup or force-refreshing the page.

企业微信截图_17237071661312
There are some issues. When I downloaded the D_X and E_X v0.21 firmware from https://esc-configurator.com/, it failed and showed an error saying 'unable to open the file'.

**Comment by stylesuxx** [comment]
Can you please post the log? If possible please also switch to english so we can understand what's going on here.
Also have you tried downloading the hex files from the release section and flashing them from disk?

**Comment by HAKRC** [comment]
> Can you please post the log? If possible please also switch to english so we can understand what's going on here. Also have you tried downloading the hex files from the release section and flashing them from disk?

kappframework-QZQKpC(1)(1)

**Comment by HAKRC** [comment]
> Can you please post the log? If possible please also switch to english so we can understand what's going on here. Also have you tried downloading the hex files from the release section and flashing them from disk?

kappframework-QZQKpC(1)(1)
kappframework-uiwirS(1)(1)

**Comment by HAKRC** [comment]
> Can you please post the log? If possible please also switch to english so we can understand what's going on here. Also have you tried downloading the hex files from the release section and flashing them from disk?

Yes，the hex files from the release section have already passed testing,but I flashed the D_X and E_X v0.21 file from https://esc-configurator.com/, it failed and showed an error saying 'Could not get file for flashing'.

**Comment by hyp0dermik-code** [comment]
You are flashing the wrong version.
This target was only added in 0.21.1-RC1, flash that version as per my instructions here https://github.com/bird-sanctuary/bluejay/pull/196#issuecomment-2285892673

**Comment by HAKRC** [comment]
> You are flashing the wrong version. This target was only added in 0.21.1-RC1, flash that version as per my instructions here [#196 (comment)](https://github.com/bird-sanctuary/bluejay/pull/196#issuecomment-2285892673)

图片1

I can only see the 0.21.0 version from https://esc-configurator.com/ ，What problem do I meet?

**Comment by HAKRC** [comment]
> You are flashing the wrong version. This target was only added in 0.21.1-RC1, flash that version as per my instructions here [#196 (comment)](https://github.com/bird-sanctuary/bluejay/pull/196#issuecomment-2285892673)

Thank you for your help.We have succeeded  to downloaded  E_X_v0.21.1-RC1 from https://esc-configurator.com/,and passed the test.

**Pull request merged**